### PR TITLE
Support privacy group in Tessera

### PR DESCRIPTION
**Pull request by namtruong**
Introduce the concept of privacy group in Tessera, add support for operations related to privacy groups such as managing privacy group data and handling transactions that are intended for a privacy group.

New APIs created include

- `/createPrivacyGroup`
- `/retrievePrivacyGroup`
- `/findPrivacyGroup`
- `/deletePrivacyGroup`

A new configuration flag is introduced to support current Besu users running Orion as its privacy manager `{ "mode": "orion"}` (default value is `"tessera"`). Once enabled, Tessera will alter its behaviour to accept requests from Besu and return responses that are complying to the existing Besu<>Orion models.

### CODE REVIEW

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java [hunk 1]
```diff
@@ -1530,8 +1531,10 @@ public void encodeForSpecificRecipientNoPsv() {
         assertThat(result.getRecipientNonce()).isEqualTo(control.getRecipientNonce());
         assertThat(result.getCipherTextNonce()).isEqualTo(control.getCipherTextNonce());
         assertThat(result.getRecipientKeys()).hasSize(1).containsExactly(recipientKey);
-        //assertThat(result.getRecipientBoxes()).isNotEqualTo(control.getRecipientBoxes());
+        // assertThat(result.getRecipientBoxes()).isNotEqualTo(control.getRecipientBoxes());
```
Comment (ricardolyn) [inline enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: non-blocking: should be removed?

File: tessera-jaxrs/transaction-jaxrs/src/test/java/com/quorum/tessera/q2t/AsyncBatchPrivacyGroupPublisherTest.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        mockExecutorFactory = mock(ExecutorFactory.class);
+        when(mockExecutorFactory.createCachedThreadPool()).thenReturn(mockExecutor);
+
+        mockCountDownLatchFactory = mock(CancellableCountDownLatchFactory.class);
+        when(mockCountDownLatchFactory.create(anyInt())).thenReturn(mockCountDownLatch);
+
+        //        when(own.getUrl()).thenReturn("http://own.com/");
+        //        when(own.getRecipientsAsMap())
+        //                .thenReturn(Map.of(recipient, "http://url1.com/", otherRecipient, "http://url2.com/"));
+        //        when(mockDiscovery.getCurrent()).thenReturn(own);
```
Comment (ricardolyn) [inline tessera-jaxrs/transaction-jaxrs/src/test/java/com/quorum/tessera/q2t/AsyncBatchPrivacyGroupPublisherTest.java]: non-blocking: do we need to keep this commented asserts?

File: tests/acceptance-test/src/test/java/config/ConfigBuilder.java [hunk 3]
```diff
@@ -236,23 +236,25 @@ public Config build() {
         config.setKeys(new KeyConfiguration());
 
         final List<KeyData> pairs =
-            keys.entrySet().stream()
-                //  .map(e -> new DirectKeyPair(e.getKey(), e.getValue()))
```
Comment (ricardolyn) [inline tests/acceptance-test/src/test/java/config/ConfigBuilder.java:240]: same here

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java (hunk 1 above)
Comment (macfarla) [inline enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: can this be deleted if not relevant?

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java [hunk 4]
```diff
@@ -2078,149 +2086,4 @@ public void withRecipientForEncodedPayloadThatHasEmptyArrayListRecipientKeys() {
 
         assertThat(payload.getRecipientKeys()).containsExactly(PublicKey.from("someKey".getBytes()));
     }
-
-    /**
-     * Do NOT change as this is a copy of the legacy payload encoder. The tests will use these logic to ensure the
-     * legacy encoder is still able to understand new encoded payload and vice versa
-     */
```
Comment (macfarla) [inline enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: is this no longer required?

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/encoder/V2PayloadEncoder.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            buffer.get(box);
+            recipientBoxes.add(box);
+        }
+
+        final long recipientNonceSize = buffer.getLong();
+        final byte[] recipientNonce = new byte[Math.toIntExact(recipientNonceSize)];
+        buffer.get(recipientNonce);
+
+        // this means there are no recipients in the payload (which we receive when we are a participant)
+        // TODO - not sure this is right
```
Comment (macfarla) [inline enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/encoder/V2PayloadEncoder.java]: what action is required to verify this?

File: tessera-core/src/test/java/com/quorum/tessera/transaction/EncodedPayloadManagerImplTest.java [hunk 6]
```diff
@@ -192,30 +197,27 @@ public void decryptBadCipherText() {
...
-            .withSenderKey(sender)
-            .withRecipientKeys(recipients)
-            .withCipherText(testPayload.getBytes())
-            .withPrivacyMode(PrivacyMode.STANDARD_PRIVATE)
-            .withAffectedContractTransactions(emptyMap())
-            .withExecHash(new byte[0])
-            .build();
-
-        when(messageHashFactory.createFromCipherText(any())).thenReturn(new MessageHash("test hash".getBytes()));
+        final List<PublicKey> recipients = List.of(singleRecipient); // list the keys multiple times
```
Comment (macfarla) [inline tessera-core/src/test/java/com/quorum/tessera/transaction/EncodedPayloadManagerImplTest.java]: here it is only once, not multiple times

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java (hunk 4 above)
Comment (namtruong) [reply enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: It was moved to a separate class https://github.com/ConsenSys/tessera/blob/fbd2cfaeecea4ba50b4d7e477f503d64b4309a67/enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/encoder/LegacyPayloadEncoder.java

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/encoder/V2PayloadEncoder.java (hunk 5 above)
Comment (namtruong) [reply enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/encoder/V2PayloadEncoder.java]: This is out of the scope for this change and can be addressed in a separate PR. The encoder was copied as-is for testing purpose.

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java (hunk 1 above)
Comment (namtruong) [reply enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: done

File: tessera-core/src/test/java/com/quorum/tessera/transaction/EncodedPayloadManagerImplTest.java (hunk 6 above)
Comment (namtruong) [reply tessera-core/src/test/java/com/quorum/tessera/transaction/EncodedPayloadManagerImplTest.java]: done

File: enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java (hunk 1 above)
Comment (namtruong) [reply enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java]: done

File: tessera-jaxrs/transaction-jaxrs/src/test/java/com/quorum/tessera/q2t/AsyncBatchPrivacyGroupPublisherTest.java (hunk 2 above)
Comment (namtruong) [reply tessera-jaxrs/transaction-jaxrs/src/test/java/com/quorum/tessera/q2t/AsyncBatchPrivacyGroupPublisherTest.java]: done

File: tests/acceptance-test/src/test/java/config/ConfigBuilder.java (hunk 3 above)
Comment (namtruong) [reply tests/acceptance-test/src/test/java/config/ConfigBuilder.java:240]: done

**Pull request merged**